File: js/src/render.js
import colors from "./colors.js";
import { boardSize } from "./constants.js";
import { tetromino } from "./tetrominos.js";

function rt(ctx, tet, rotation = 0, x = 0, y = 0) {
  const unit = boardSize.unit;
  const t = tetromino(tet, rotation);
  const size = t.length;

  t.forEach((row, r) => {
    row.forEach((col, c) => {
      const h = size - r + y - 1;
      if (col) {
        ctx.fillStyle = colors[col];
        ctx.strokeStyle = colors.bg;
        ctx.lineWidth = 3;
        ctx.fillRect((c + x) * unit, h * unit, unit, unit);
        ctx.strokeRect((c + x) * unit, h * unit, unit, unit);
      }
    });
  });
}

export function renderTetromino(ctx, tetromino) {
  ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);
  rt(ctx, tetromino);
}

export function render(ctx, board, current, rotation, x, y) {
  const unit = boardSize.unit;
  ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height);

  board.forEach((row, r) => {
    row.forEach((col, c) => {
      if (col) {
        ctx.fillStyle = colors[col];
        ctx.strokeStyle = colors.bg;
        ctx.lineWidth = 3;
        ctx.fillRect(c * unit, r * unit, unit, unit);
        ctx.strokeRect(c * unit, r * unit, unit, unit);
      }
    });
  });

  rt(ctx, current, rotation, x, y);
}

File: python/tetris/render.py
import curses
from tetris.context import context
from tetris.tetrominos import tetromino

colors = (0, 1, 2, 3, 3, 4, 4, 2)


def render_tetromino(type, rotation, x, y):
    t = tetromino(type, rotation)
    size = len(t)

    for r, row in enumerate(t):
        for c, col in enumerate(row):
            h = y - size + r + 1
            l = c * 2 + x
            if col != 0:
                context["screen"].addstr(h, l, "██", curses.color_pair(colors[type]))


def rt_on_board(type, rotation, x, y):
    render_tetromino(type, rotation, context["x0"] + x * 2, context["y0"] - y)


def render_board(board):
    for r, row in enumerate(board):
        for c, col in enumerate(row):
            if col != 0:
                context["screen"].addstr(
                    context["y0"] - r,
                    c * 2 + context["x0"],
                    "██",
                    curses.color_pair(colors[col]),
                )
